### DISCUSSION TITLE
Reinitialize in a Multi-Scope Setup

### DISCUSSION THREAD

**Discussion by BluePalmTree** [Mimetis/Dotmim.Sync#1047] [Q&A] [answered]
Hi,

we have a Multi-Scope Setup, where the first Scope is the ConfigScope and others include i.e. ProductScope. Now the Product-Table (in ProductScope) has a foreign Key to a Table in the ConfigScope. The ConfigScope gets synced before the ProductScope.
The Problem is, if the ConfigScope is outdated we do this:

```
localOrchestrator.OnOutdated(async args =>
{
	OnLogAdded(args, null, DbSyncMessageType.Warning);

	var eventargs = new DbSyncScopeOutdatedEventArgs();
	ScopeOutdated?.Invoke(this, eventargs);

	await Task.WhenAll(eventargs.ReinitializeAfterOutdated.Task);

	args.Action = OutdatedAction.ReinitializeWithUpload;
});
```

The Problem is we get this error `The DELETE statement conflicted with the REFERENCE constraint ...` which is understandable. My Question is what would be the best approach for this case? What I can think of:

1. Create one Scope with all the tables in it and reinitialize it when **any** Scope is outdated
2. Disable foreign key constraints for the reinitialize sync somehow

**Comment by BluePalmTree** [comment]
I tried option 2 even just setting the `DisableConstraintsOnApplyChanges` in the setup since I know the scope is outdated. But I still get the Foreign Key error on the delete statement.

**Reply by Mimetis** [reply]
I think you have another table outside the scope you are trying to reinitialize, that is conflicting.

You should try to manually reinitialize all the tables, by dropping the tracking table content (if not using change tracking) and manually dropping the content of the tables

**Reply by BluePalmTree** [reply]
I don't think so, but here is a clearer explanation. We have the `ConfigScope` with multiple tables one of them is the `SalesTypes`. Then we have the `ProductScope` with one table `Products`. Now if I try to `ReinitializeWithUpload` the `ConfigScope` (which is synced before the `ProductScope`) I get the error: T`he DELETE statement conflicted with the REFERENCE constraint "FK_Products_SalesTypes_SalesTypeId". The conflict occurred in database "AdventureWorks", table "dbo.Products", column 'SalesTypeId'.`. I set `DisableConstraintsOnApplyChanges = true` for the `ConfigScope`.

We're using the ChangeTracking-Provider. Is there a way to do this via Dotmim.Sync or do we have to manually do the dropping of the content? If we manually drop the content wouldn't Dotmim try to sync the deleted rows to the server?

**Reply by Mimetis** [reply, marked as answer]
> We're using the ChangeTracking-Provider. Is there a way to do this via Dotmim.Sync or do we have to manually do the dropping of the content? If we manually drop the content wouldn't Dotmim try to sync the deleted rows to the server?

Yes.
If you want to do it manually, you need to:
- `ALTER TABLE TABLE1 DISABLE CHANGE_TRACKING;`
- `DELETE FROM TABLE1`
- `ALTER TABLE TABLE1 ENABLE CHANGE_TRACKING;`

**Reply by BluePalmTree** [reply]
Thank you for the info. One last question is there a way to let the client try to upload unsynced data without the reinitialize?

**Reply by Mimetis** [reply]
if outdated, I dont think so

**Comment by BluePalmTree** [comment]
> I think you have another table outside the scope you are trying to reinitialize, that is conflicting.

The issue is, that in this relatioship the Primary Key Base Table is the SalesTypes-Table and the Foreign Key Base Table is the Products-Table, the relationship is associated with the Products-Table. So if `DisableConstraintsOnApplyChanges = true` is set it doesn't affect this relationship because the Products-Table is in a different scope.
